src/lib/cmd.rs

```diff
@@ -1,64 +1,82 @@
 use std::fs::File;
 use std::process::{Command, Stdio};
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use crate::lib::job;
 
 use super::config;
 
-fn remote_cmd(args: &mut Vec<&str>, output: &mut File) -> Result<(), String> {
+async fn remote_cmd(args: &mut Vec<&str>, output: &mut File, curr_job: &Arc<RwLock<job::JobInfo>>) -> Result<(), String> {
     // TODO(azhng): let's finger cross this works.
     let output_file = output.try_clone().unwrap();
     let error_file = output.try_clone().unwrap();
-    let cmd = Command::new("ssh")
+    let cmd_process = Command::new("ssh")
         .arg(config::SEXXI_REMOTE_HOST)
-        .args(args)
+        .args(args.clone())
         .stdout(Stdio::from(output_file))
         .stderr(Stdio::from(error_file))
-        .output()
-        .expect("Ok");
+        .spawn();
+
+    match cmd_process {
+        Ok(child) => {
+            info!("Server process with pid {} is running command `{:?}`", child.id(), args);
+            {
+                let mut job_info = curr_job.write().await;
+                job_info.pid = child.id();
+            }
+
+            let result = child.wait_with_output().expect("Ok");
+            if !result.status.success() {
+                return Err(format!("remote command failed: {}", result.status));
+            }
+        },
+        Err(e) => {
+            error!("Command `{:?}` failed, server process didn't start: {}", args, e);
+            std::process::exit(1);
+        },
 
-    if !cmd.status.success() {
-        return Err(format!("remote command failed: {}", cmd.status));
     }
 
     Ok(())
 }
 
-fn remote_git_cmd(args: &mut Vec<&str>, output: &mut File) -> Result<(), String> {
+async fn remote_git_cmd(args: &mut Vec<&str>, output: &mut File, curr_job: &Arc<RwLock<job::JobInfo>>) -> Result<(), String> {
```

Comment:
any particular reason that all the `remote_*` functions are converted into async?

Reply:
Yes, otherwise compiler yells at me for using `.await`; an async function can only be used in an async function

Reply:
Ok I see the issue comes from `let mut job_info = curr_job.write().await;` on L24. Though I think addressing the comments in the `jobr.rs` first will make this look at lot cleaner. I guess we can come back to this once this PR is unWIP'd